File: ice-core/src/main/java/com/ice/core/relation/All.java
package com.ice.core.relation;

import com.ice.common.enums.NodeRunStateEnum;
import com.ice.core.base.BaseNode;
import com.ice.core.base.BaseRelation;
import com.ice.core.context.IceContext;
import com.ice.core.utils.IceLinkedList;

/**
 * @author waitmoon
 * relation ALL
 * all child will execute
 * have TRUE--TRUE
 * without TRUE have FALSE--FALSE
 * without chilren--NONE
 * all NONE--NONE
 */
public final class All extends BaseRelation {
    /*
     * process relation all
     */
    @Override
    protected NodeRunStateEnum processNode(IceContext ctx) {
        IceLinkedList<BaseNode> children = this.getChildren();
        if (children == null || children.isEmpty()) {
            return NodeRunStateEnum.NONE;
        }
        if (children.getSize() == 1) {
            BaseNode node = children.get(0);
            if (node == null) {
                return NodeRunStateEnum.NONE;
            }
            return node.process(ctx);
        }
        boolean hasTrue = false;
        boolean hasFalse = false;
        for (IceLinkedList.Node<BaseNode> listNode = children.getFirst(); listNode != null; listNode = listNode.next) {
            BaseNode node = listNode.item;
            if (node != null) {
                NodeRunStateEnum stateEnum = node.process(ctx);
                if (!hasTrue) {
                    hasTrue = stateEnum == NodeRunStateEnum.TRUE;
                }
                if (!hasFalse) {
                    hasFalse = stateEnum == NodeRunStateEnum.FALSE;
                }
            }
        }
        if (hasTrue) {
            return NodeRunStateEnum.TRUE;
        }
        if (hasFalse) {
            return NodeRunStateEnum.FALSE;
        }
        return NodeRunStateEnum.NONE;
    }
}

File: sdks/go/relation/all.go
package relation

import (
	stdctx "context"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
	"github.com/zjn-zjn/ice/sdks/go/enum"
	"github.com/zjn-zjn/ice/sdks/go/node"
)

// All is a relation node that executes all children.
// - Has TRUE -> TRUE
// - Without TRUE, has FALSE -> FALSE
// - Without children -> NONE
// - All NONE -> NONE
type All struct {
	node.Relation
}

// NewAll creates a new All relation node.
func NewAll() *All {
	return &All{
		Relation: *node.NewRelation(),
	}
}

// Process implements the Node interface.
func (a *All) Process(ctx stdctx.Context, iceCtx *icecontext.Context) enum.RunState {
	return node.ProcessWithBase(ctx, &a.Base, iceCtx, a.processNode)
}

func (a *All) processNode(ctx stdctx.Context, iceCtx *icecontext.Context) enum.RunState {
	if a.Children == nil || a.Children.IsEmpty() {
		return enum.NONE
	}

	hasTrue := false
	hasFalse := false
	for listNode := a.Children.First(); listNode != nil; listNode = listNode.Next {
		n := listNode.Item
		if n != nil {
			state := n.Process(ctx, iceCtx)
			if !hasTrue && state == enum.TRUE {
				hasTrue = true
			}
			if !hasFalse && state == enum.FALSE {
				hasFalse = true
			}
		}
	}

	if hasTrue {
		return enum.TRUE
	}
	if hasFalse {
		return enum.FALSE
	}
	return enum.NONE
}
